### DISCUSSION TITLE
AttributeError exception with Pythonnet upgrade to 3.0.1

### DISCUSSION THREAD

**Discussion by BrelKsiba** [pythonnet/pythonnet#2006] [Q&A]
Pythonnet 3.0.1
Python 3.10
Windows 10

I have some scripts running without problems with old version pythonnet 2.5.2. When we upgrade to 3.0.1 and Python to 3.10 we are getting an attribute error exception: 

```
AttributeError: 'IBaseValue' object has no attribute 'Value'. 
```

The code is simple:

**In python side...**

```python
MAPort.Read(l_name).Value
```

**Then in C# side Read function is defined as follow :** 

```csharp
public IBaseValue Read(string variableName)
{
// some code...
     case DataType.eBOOLEAN:
         IBooleanValue bv = element.data as IBooleanValue;
         val = get_my_internal_value();
         bv.Value = Convert.ToBoolean(val);
         return element.data;
}
```

In this case `element.data` is `IBooleanValue` which is inherited from `IBaseValue`.

Any help or advice is welcome.

Thx!

**Comment by filmor** [comment]
My guess would be that you are running into the changed behaviour from https://github.com/pythonnet/pythonnet/pull/1240 here. Your `IBaseValue` probably doesn't have the `Value` accessor, right? You can try casting explicitly using `IBooleanValue(MAPort.Read(l_name)).Value`.

**Comment by BrelKsiba** [comment]
Thank you. The explicit cast as you proposed worked for me. 
The Read function have several type of returns (depending on datatypes). This means that we have to check the data type each time and  then do the corresponding explicit cast. This makes the Python script more complex. 
I will read all the thread you mentioned and see if there any other alternatives. 
Thank you again for your prompt support!

**Comment by lostmsu** [comment]
@BrelKsiba there's also `__implementation__` and `__raw_implementation__`

**Reply by drblarg** [reply]
@lostmsu I raw across the same issue and I am looking for the best solution from the python side (.NET side is from a 3rd party). Is there a more detailed comparison (pros/cons) between explicit casting, `__implementation__`, and `__raw_implementation__`?

**Reply by lostmsu** [reply]
Not sure about explicit casting, but `__implementation__` will call codecs registered for the actual type if any, and `__raw_implementation__` won't.
